File: indexer.py
# Eratosthenes - an APT repository browser and indexer, designed for Vanilla OS.
# Copyright (C) 2023-2025 Vanilla OS <https://vanillaos.org>

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import gzip
import lzma
import logging
import sqlite3
import requests
from io import BytesIO

from config import DB_PATH, BRANCHES, REPO_COMPONENTS, ARCHS

logger = logging.getLogger("AptIndexer")


class AptIndexer:
    def __init__(self):
        self.db = sqlite3.connect(DB_PATH)
        self.cursor = self.db.cursor()
        self.create_schema()

    def create_schema(self):
        self.cursor.execute(
            """
            CREATE TABLE IF NOT EXISTS packages (
                id INTEGER PRIMARY KEY,
                name TEXT,
                version TEXT,
                description TEXT,
                section TEXT,
                homepage TEXT,
                maintainer TEXT,
                depends TEXT,
                recommends TEXT,
                suggests TEXT,
                conflicts TEXT,
                replaces TEXT,
                provides TEXT,
                filename TEXT,
                branch TEXT,
                arch TEXT
            )
        """
        )
        # Ensure 'arch' column exists in case an old table persists
        self.cursor.execute("PRAGMA table_info(packages)")
        columns = [column[1] for column in self.cursor.fetchall()]
        if "arch" not in columns:
            self.cursor.execute("ALTER TABLE packages ADD COLUMN arch TEXT")
        self.db.commit()

    def cleanup(self):
        self.cursor.execute("DROP TABLE IF EXISTS packages")
        self.db.commit()
        self.create_schema()

    def index(self):
        for branch, base_url in BRANCHES.items():
            for component in REPO_COMPONENTS:
                for arch in ARCHS:
                    # Replace first @ with component and second @ with binary-<arch>
                    url = base_url.replace("@", component, 1).replace("@", f"binary-{arch}", 1)
                    
                    logger.info(
                        f"Indexing branch '{branch}' component '{component}' arch '{arch}' at {url}"
                    )
                    
                    response = None
                    # Try Packages, then Packages.gz, then Packages.xz
                    for ext in ["", ".gz", ".xz"]:
                        try:
                            fetch_url = url + ext
                            r = requests.get(fetch_url, timeout=10)
                            if r.status_code == 200:
                                if ext == ".gz":
                                    content = gzip.decompress(r.content).decode("utf-8")
                                elif ext == ".xz":
                                    content = lzma.decompress(r.content).decode("utf-8")
                                else:
                                    content = r.text
                                response = content
                                break
                        except Exception as e:
                            logger.error(f"Error fetching {fetch_url}: {e}")
                            continue
                    
                    if not response:
                        logger.warning(f"Could not find Packages for {branch}/{component}/{arch}")
                        continue

                    data = response.split("\n\n")
                    values = []
                    for pkg_raw in data:
                        if not pkg_raw.strip():
                            continue
                        lines = pkg_raw.split("\n")
                        name = version = description = section = homepage = maintainer = (
                            None
                        )
                        depends = recommends = suggests = conflicts = replaces = (
                            provides
                        ) = filename = None

                        for line in lines:
                            if line.startswith("Package: "):
                                name = line.split(": ", 1)[1]
                            elif line.startswith("Version: "):
                                version = line.split(": ", 1)[1]
                            elif line.startswith("Description: "):
                                description = line.split(": ", 1)[1]
                            elif line.startswith("Section: "):
                                section = line.split(": ", 1)[1]
                            elif line.startswith("Homepage: "):
                                homepage = line.split(": ", 1)[1]
                            elif line.startswith("Maintainer: "):
                                maintainer = line.split(": ", 1)[1]
                            elif line.startswith("Depends: "):
                                depends = line.split(": ", 1)[1]
                            elif line.startswith("Recommends: "):
                                recommends = line.split(": ", 1)[1]
                            elif line.startswith("Suggests: "):
                                suggests = line.split(": ", 1)[1]
                            elif line.startswith("Conflicts: "):
                                conflicts = line.split(": ", 1)[1]
                            elif line.startswith("Replaces: "):
                                replaces = line.split(": ", 1)[1]
                            elif line.startswith("Provides: "):
                                provides = line.split(": ", 1)[1]
                            elif line.startswith("Filename: "):
                                filename = line.split(": ", 1)[1]

                        if name:
                            values.append(
                                (
                                    name,
                                    version,
                                    description,
                                    section,
                                    homepage,
                                    maintainer,
                                    depends,
                                    recommends,
                                    suggests,
                                    conflicts,
                                    replaces,
                                    provides,
                                    filename,
                                    branch,
                                    arch,
                                )
                            )
                    
                    if values:
                        self.cursor.executemany(
                            """
                            INSERT INTO packages (
                                name, version, description, section, homepage, maintainer,
                                depends, recommends, suggests, conflicts, replaces, provides,
                                filename, branch, arch
                            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                        """,
                            values,
                        )
                        self.db.commit()
                        logger.info(
                            f"Finished indexing {len(values)} packages for branch '{branch}' component '{component}' arch '{arch}'."
                        )

File: internal/indexer/indexer.go
package indexer

import (
	"bufio"
	"compress/gzip"
	"context"
	"fmt"
	"io"
	"net/http"
	"strings"

	"github.com/ulikunitz/xz"
	"github.com/vanilla-os/eratosthenes/internal/config"
	"github.com/vanilla-os/eratosthenes/internal/db"
	"github.com/vanilla-os/eratosthenes/internal/models"
)

type Logger interface {
	Info(msg string)
	Infof(format string, args ...any)
	Error(msg string)
	Errorf(format string, args ...any)
}

type Indexer struct {
	logger Logger
}

func NewIndexer(l Logger) *Indexer {
	return &Indexer{logger: l}
}

func (i *Indexer) Cleanup() {
	i.logger.Info("Note: Cleanup for Slipstream involves overwriting keys. Obsolete keys might persist until compaction/expiry.")
}

func (i *Indexer) Index() {
	for branch, baseUrl := range config.Instance.Branches {
		for _, component := range config.Instance.RepoComponents {
			for _, arch := range config.Instance.Archs {
				url := strings.Replace(baseUrl, "@", component, 1)
				url = strings.Replace(url, "@", fmt.Sprintf("binary-%s", arch), 1)

				i.logger.Infof("Indexing branch '%s' component '%s' arch '%s' at %s", branch, component, arch, url)

				var content string
				var err error

				for _, ext := range []string{"", ".gz", ".xz"} {
					content, err = i.fetch(url + ext)
					if err == nil {
						break
					}
					if ext == ".xz" {
						i.logger.Infof("Error fetching %s: %v", url+ext, err)
					}
				}

				if content == "" {
					i.logger.Infof("Could not find Packages for %s/%s/%s", branch, component, arch)
					continue
				}

				i.parseAndInsert(content, branch, arch)
			}
		}
	}
}

func (i *Indexer) fetch(url string) (string, error) {
	resp, err := http.Get(url)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return "", fmt.Errorf("status code %d", resp.StatusCode)
	}

	var reader io.Reader = resp.Body

	if strings.HasSuffix(url, ".gz") {
		gzReader, err := gzip.NewReader(resp.Body)
		if err != nil {
			return "", err
		}
		defer gzReader.Close()
		reader = gzReader
	} else if strings.HasSuffix(url, ".xz") {
		xzReader, err := xz.NewReader(resp.Body)
		if err != nil {
			return "", err
		}
		reader = xzReader
	}

	bytes, err := io.ReadAll(reader)
	if err != nil {
		return "", err
	}

	return string(bytes), nil
}

func (i *Indexer) parseAndInsert(content, branch, arch string) {
	chunks := strings.Split(content, "\n\n")
	ctx := context.Background()

	tx, err := db.DB.Begin()
	if err != nil {
		i.logger.Errorf("Error starting transaction: %v", err)
		return
	}

	count := 0
	for _, chunk := range chunks {
		if strings.TrimSpace(chunk) == "" {
			continue
		}

		pkgMap := make(map[string]string)
		scanner := bufio.NewScanner(strings.NewReader(chunk))
		var currentKey string

		for scanner.Scan() {
			line := scanner.Text()
			if strings.HasPrefix(line, " ") {
				if currentKey != "" {
					pkgMap[currentKey] += "\n" + strings.TrimSpace(line)
				}
				continue
			}

			parts := strings.SplitN(line, ": ", 2)
			if len(parts) == 2 {
				currentKey = parts[0]
				pkgMap[currentKey] = parts[1]
			}
		}

		if _, ok := pkgMap["Package"]; !ok {
			continue
		}

		pkg := models.Package{
			Name:        pkgMap["Package"],
			Version:     pkgMap["Version"],
			Description: pkgMap["Description"],
			Section:     pkgMap["Section"],
			Homepage:    pkgMap["Homepage"],
			Maintainer:  pkgMap["Maintainer"],
			Depends:     pkgMap["Depends"],
			Recommends:  pkgMap["Recommends"],
			Suggests:    pkgMap["Suggests"],
			Conflicts:   pkgMap["Conflicts"],
			Replaces:    pkgMap["Replaces"],
			Provides:    pkgMap["Provides"],
			Filename:    pkgMap["Filename"],
			Branch:      branch,
			Arch:        arch,
		}

		key := fmt.Sprintf("%s|%s|%s", branch, arch, pkg.Name)

		if err := tx.Put(ctx, key, pkg, 0); err != nil {
			i.logger.Errorf("Error putting package %s: %v", pkg.Name, err)
		} else {
			count++
		}

		if count%1000 == 0 {
			if err := tx.Commit(ctx); err != nil {
				i.logger.Errorf("Error committing batch: %v", err)
			}
			tx, _ = db.DB.Begin()
		}
	}

	if err := tx.Commit(ctx); err != nil {
		i.logger.Errorf("Error committing final transaction: %v", err)
	}

	i.logger.Infof("Finished indexing %d packages for branch '%s' arch '%s'", count, branch, arch)
}
